drivers/stepper/adi/tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,608 @@
+}
+
+#endif
+
+static int tmc5041_init(const struct device *dev)
+{
+	LOG_DBG("TMC5041 stepper motor controller %s initialized", dev->name);
+	struct tmc5041_data *data = dev->data;
+	const struct tmc5041_config *config = dev->config;
+
+	(void)k_mutex_init(&data->mutex);
+	if (!spi_is_ready_dt(&config->spi)) {
+		LOG_ERR("SPI bus is not ready");
+		return -ENODEV;
+	}
+
+	/* Read GSTAT register values to clear any errors SPI Datagram. */
+	uint32_t gstat_value;
+	(void)tmc5041_read(dev, TMC5041_GSTAT, &gstat_value);
+
+	/* Init non motor-index specific registers here. */
+	LOG_DBG("GCONF: %d", config->gconf);
+	(void)tmc5041_write(dev, TMC5041_GCONF, config->gconf);
+	LOG_DBG("Device %s initialized", dev->name);
+
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+	(void)activate_gpio_interrupt(dev);
+#endif
+	return 0;
+}
+
+static int tmc5041_stepper_init(const struct device *dev)
+{
+	const struct tmc5041_stepper_config *stepper_config = dev->config;
+	struct tmc5041_stepper_data *data = dev->data;
+
+	LOG_DBG("Controller: %s, Stepper: %s", stepper_config->controller->name, dev->name);
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	(void)tmc5041_write(stepper_config->controller, TMC5041_SWMODE(stepper_config->index), 0);
+	(void)k_work_init_delayable(&data->stallguard_dwork, stallguard_work_handler);
+	LOG_DBG("Setting stall guard to %d", stepper_config->stallguard_threshold);
+	int32_t stall_guard_threshold = (int32_t)stepper_config->stallguard_threshold;
+	(void)tmc5041_write(stepper_config->controller, TMC5041_COOLCONF(stepper_config->index),
+			    stall_guard_threshold << TMC5041_COOLCONF_SG2_THRESHOLD_VALUE_SHIFT);
+	stallguard_enable(dev, true);
+#if STEPPER_LOG_LEVEL >= LOG_LEVEL_DBG
+	(void)tmc5041_read(stepper_config->controller, TMC5041_COOLCONF(stepper_config->index),
+			   &stall_guard_threshold);
+	LOG_DBG("Stall guard set to %d",
+		stall_guard_threshold >> TMC5041_COOLCONF_SG2_THRESHOLD_VALUE_SHIFT);
+#endif
+#else
+	ARG_UNUSED(data);
+#endif
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+	(void)tmc5041_stepper_set_ramp(dev, &stepper_config->default_ramp_config);
+#endif
+	(void)tmc5041_stepper_set_micro_step_res(dev, stepper_config->default_micro_step_res);
+	return 0;
+}
+
+#define TMC5041_STEPPER_CONFIG_DEFINE(child)							\
+	static const struct tmc5041_stepper_config tmc5041_stepper_config_##child = {		\
+		.controller = DEVICE_DT_GET(DT_PARENT(child)),					\
+		.default_micro_step_res = DT_PROP_OR(child, micro_step_res, 256),		\
+		.index = DT_REG_ADDR(child),							\
+		IF_ENABLED(CONFIG_STEPPER_ADI_TMC_STALLGUARD,					\
+		(.stallguard_threshold = DT_PROP_OR(child, stallguard_threshold, 0),))		\
+		IF_ENABLED(CONFIG_STEPPER_ADI_TMC_RAMP_GEN,					\
+		(.default_ramp_config = TMC_RAMP_DT_SPEC_GET(child))) };
+
+#define TMC5041_STEPPER_DATA_DEFINE(child)							\
+	static struct tmc5041_stepper_data tmc5041_stepper_data_##child = {			\
+		IF_ENABLED(CONFIG_STEPPER_ADI_TMC_STALLGUARD,					\
+			(.stepper = DEVICE_DT_GET(child),))};
```

Comment:
shouldn't this be in the config structure?

Reply:
i need access to device* in k_work_handler(_stallguard_handler_) to enable/disable stallguard for that particular stepper motor

Reply:
usually this is done by having a pointer to the parent device in the data struct, and then getting the config through it :)

```
struct foo_dev_config {
        ...
};

struct foo_dev_data {
        const struct device *dev;
}

void foo(struct foo_dev_data *data)
{
        const struct device dev = data->dev;
        const struct foo_dev_config config = dev.>config;
        ...
}
```

Reply:
I am doing the same as well, i need to access the the config of that device.
```
#define TMC5041_STEPPER_DATA_DEFINE(child)							\
	static struct tmc5041_stepper_data tmc5041_stepper_data_##child = {			\
		.stepper = DEVICE_DT_GET(child),};

static void stallguard_work_handler(struct k_work *work)
{
	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
	struct tmc5041_stepper_data *stepper_data =
		CONTAINER_OF(dwork, struct tmc5041_stepper_data, stallguard_dwork);
	int err;
	const struct tmc5041_stepper_config *stepper_config = stepper_data->stepper->config;

	LOG_DBG("Stallguard work handler");
#if CONFIG_STEPPER_LOG_LEVEL >= LOG_LEVEL_DBG
	int32_t sg_result;

	err = tmc5041_read(stepper_config->controller,
			   TMC5041_DRVSTATUS(stepper_config->index), &sg_result);
	if (err) {
		LOG_ERR("Failed to read DRVSTATUS register");
		return;
	}
	LOG_DBG("Detecting Load for motor %s %ld %ld", stepper_data->stepper->name,
		sg_result & TMC5041_DRV_STATUS_SG_RESULT_MASK,
		(sg_result & TMC5041_DRV_STATUS_SG_STATUS_MASK) >>
			TMC5041_DRV_STATUS_SG_STATUS_SHIFT);
#endif

	err = stallguard_enable(stepper_data->stepper, true);
	if (err == -EAGAIN) {
		LOG_ERR("Trying again to enable stallguard");

		k_work_reschedule(dwork, K_MSEC(stepper_config->stallguard_delay_in_ms));
	}
	if (err == -EIO) {
		LOG_ERR("Failed to enable stallguard becaoause of I/O error");
		return;
	}
}

```

Reply:
The struct tmc5041_stepper_data is not const, but .stepper is, you are storing immutable data in a mutable data structure (RAM)